core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala

```diff
@@ -0,0 +1,130 @@
+  }
+
+  def apply(clientInstanceId: String, clientId: String, softwareName: String,
+            softwareVersion: String, clientHostAddress: String, clientPort: String): ClientMetricsMetadata = {
+    val instance = new ClientMetricsMetadata
+    instance.init(clientInstanceId, clientId, softwareName, softwareVersion, clientHostAddress, clientPort)
+    instance
+  }
+
+  /**
+   * Parses the client matching patterns and builds a map with entries that has
+   * (PatternName, PatternValue) as the entries.
+   *  Ex: "VERSION=1.2.3" would be converted to a map entry of (Version, 1.2.3)
+   *
+   *  NOTES:
+   *  1. Client match pattern splits the input into two parts separated by first
+   *     occurrence of the character '='
+   *  2. '*' is considered as invalid client match pattern
+   * @param patterns List of client matching pattern strings
+   * @return map of client matching pattern entries
+   */
+  def parseMatchingPatterns(patterns: List[String]) : Map[String, String] = {
+    val patternsMap = mutable.Map[String, String]()
+    if (patterns != null) {
+      patterns.foreach(x => {
+        val nameValuePair = x.split("=", 2).map(x => x.trim)
+        if (nameValuePair.size == 2 && isValidParam(nameValuePair(0)) && validRegExPattern(nameValuePair(1))) {
+          patternsMap += (nameValuePair(0) -> nameValuePair(1))
+        } else {
+          throw new InvalidConfigurationException("Illegal client matching pattern: " + x)
+        }
+      })
+    }
+    patternsMap.toMap
+  }
+
+  private def validRegExPattern(inputPattern :String): Boolean = {
+    try {
+      Pattern.compile(inputPattern)
+      true
+    } catch {
+      case e: PatternSyntaxException =>
+        false
+    }
+  }
+
+}
+
+class ClientMetricsMetadata {
+  var attributesMap: mutable.Map[String, String] = scala.collection.mutable.Map[String, String]()
+
+  private def init(clientInstanceId: String,
+                   clientId: String,
+                   softwareName: String,
+                   softwareVersion: String,
+                   clientHostAddress: String,
+                   clientPort: String): Unit = {
+    attributesMap(CLIENT_INSTANCE_ID) = clientInstanceId
+    attributesMap(CLIENT_ID) = clientId
+    attributesMap(CLIENT_SOFTWARE_NAME) = softwareName
+    attributesMap(CLIENT_SOFTWARE_VERSION) = softwareVersion
+    attributesMap(CLIENT_SOURCE_ADDRESS) = clientHostAddress
+    attributesMap(CLIENT_SOURCE_PORT) = clientPort
+  }
+  def getClientId: Option[String] = attributesMap.get(CLIENT_ID)
+
+  def isMatched(patterns: Map[String, String]) : Boolean = {
+    // Empty pattern or missing pattern still considered as a match
+    if (patterns == null || patterns.isEmpty) {
+      true
+    } else {
+      matchPatterns(patterns)
+    }
+  }
+
+  private def matchPatterns(matchingPatterns: Map[String, String]) : Boolean = {
+    try {
+      matchingPatterns.foreach {
+      case (k, v) =>
+        val attribute = attributesMap.getOrElse(k, null)
```

Comment:
invalidConfiguraitonException is not needed, if you use pattern matching, you can do:

```
attributesMap.get(k) match {
case Some(attr) if v.r.anchored.findFirstIn(attribute).isDefined => 
          true
        case _ => 
          info(s"Unable to find the matching client...")
          false
```

Reply:
ah I see, you can do matchingPattern.find() which returns on the first true (so instead of returning false for `case _` you can return true.  S

Reply:
I have simplified the code, see if it makes more sense now or still you want to go with your suggestion.